File: engine/index_gc.go
// Copyright (c) 2026 Kiruba Sankar Swaminathan
//
// This source code is licensed under the MIT license found in the
// LICENSE file in the root of this source tree.

package engine

import (
	"math"
	"sync/atomic"

	"turnstone/engine/hashindex"
)

const defaultIndexFragmentationRatio = 3.0

// IndexGCReader captures one active transaction snapshot for index pruning.
type IndexGCReader struct {
	Snapshot Snapshot
	MyXid    uint64
	Update   bool
}

// IndexGCContext is a point-in-time view of which index versions must be kept.
type IndexGCContext struct {
	Readers    []IndexGCReader
	ActiveXids map[uint64]struct{}
	LogFloor   int64
	Clog       func(uint64) TxStatus
	Visible    func(uint64, Snapshot) bool
}

// BuildIndexGCContext snapshots active transactions and the log scan floor.
func (db *DB) BuildIndexGCContext() IndexGCContext {
	db.activeTxnsMu.Lock()
	txs := make([]*Transaction, 0, len(db.activeTxns))
	for tx := range db.activeTxns {
		txs = append(txs, tx)
	}
	db.activeTxnsMu.Unlock()

	db.txMu.Lock()
	activeXids := make(map[uint64]struct{}, len(db.activeXids))
	for xid := range db.activeXids {
		activeXids[xid] = struct{}{}
	}
	db.txMu.Unlock()

	readers := make([]IndexGCReader, 0, len(txs))
	for _, tx := range txs {
		readers = append(readers, IndexGCReader{
			Snapshot: tx.snapshot,
			MyXid:    tx.xid,
			Update:   tx.update,
		})
	}

	return IndexGCContext{
		Readers:    readers,
		ActiveXids: activeXids,
		LogFloor:   atomic.LoadInt64(&db.scanFloor),
		Clog:       db.clogStatus,
		Visible:    db.isVisible,
	}
}

// FilterVersions returns the version chain to retain for key (newest first).
func (ctx IndexGCContext) FilterVersions(_ []byte, chain []indexVersion) []indexVersion {
	if len(chain) == 0 {
		return nil
	}
	mask := ctx.keepMask(chain)
	out := make([]indexVersion, 0, len(chain))
	for i, v := range chain {
		if mask[i] {
			out = append(out, v)
		}
	}
	return out
}

func (ctx IndexGCContext) keepMask(chain []indexVersion) []bool {
	n := len(chain)
	keep := make([]bool, n)

	if len(ctx.Readers) == 0 {
		for i, v := range chain {
			keep[i] = true
			if ctx.Clog(v.xmin) == TxCommitted {
				break
			}
		}
	} else {
		for _, r := range ctx.Readers {
			for i, v := range chain {
				keep[i] = true
				if r.Update && v.xmin == r.MyXid {
					break
				}
				if ctx.Visible(v.xmin, r.Snapshot) {
					break
				}
			}
		}
	}

	for i, v := range chain {
		if _, ok := ctx.ActiveXids[v.xmin]; ok {
			keep[i] = true
		}
		if ctx.LogFloor > 0 && v.offset < ctx.LogFloor {
			keep[i] = false
		}
	}
	return keep
}

// IndexCompactResult reports arena shrink from a compact pass.
type IndexCompactResult struct {
	ShardsCompacted int
	ArenaBefore     uint64
	ArenaAfter      uint64
}

// CompactIndex rewrites every shard, dropping unneeded versions and reclaiming arena space.
func (db *DB) CompactIndex(ctx IndexGCContext) (IndexCompactResult, error) {
	if db.index == nil || db.index.hash == nil {
		return IndexCompactResult{}, nil
	}

	statsBefore := db.index.hash.Stats()
	var before uint64
	for i := range statsBefore.Shards {
		before += statsBefore.Shards[i].ArenaUsed
	}

	statsAfter, err := db.index.hash.CompactAll(db.indexVersionFilter(ctx))
	if err != nil {
		return IndexCompactResult{}, err
	}

	var after uint64
	for i := range statsAfter.Shards {
		after += statsAfter.Shards[i].ArenaUsed
	}

	return IndexCompactResult{
		ShardsCompacted: numHashShards(statsBefore),
		ArenaBefore:     before,
		ArenaAfter:      after,
	}, nil
}

func numHashShards(stats hashindex.IndexStats) int {
	n := 0
	for i := range stats.Shards {
		if stats.Shards[i].KeyCount > 0 {
			n++
		}
	}
	return n
}

// MaybeCompactIndex compacts shards whose arena is substantially larger than live data.
func (db *DB) MaybeCompactIndex() (IndexCompactResult, error) {
	if db.index == nil || db.index.hash == nil {
		return IndexCompactResult{}, nil
	}

	ratio := db.indexFragmentationRatio
	if ratio <= 0 {
		ratio = defaultIndexFragmentationRatio
	}

	statsBefore := db.index.hash.Stats()
	ctx := db.BuildIndexGCContext()
	filter := db.indexVersionFilter(ctx)

	var before, after uint64
	compacted := 0
	for i := range statsBefore.Shards {
		st := statsBefore.Shards[i]
		before += st.ArenaUsed
		if st.KeyCount == 0 || st.LiveBytes == 0 {
			after += st.ArenaUsed
			continue
		}
		if float64(st.ArenaUsed) <= float64(st.LiveBytes)*ratio {
			after += st.ArenaUsed
			continue
		}
		afterStats, err := db.index.hash.CompactShard(i, filter)
		if err != nil {
			return IndexCompactResult{}, err
		}
		compacted++
		after += afterStats.ArenaUsed
	}

	if compacted == 0 {
		return IndexCompactResult{}, nil
	}
	return IndexCompactResult{
		ShardsCompacted: compacted,
		ArenaBefore:     before,
		ArenaAfter:      after,
	}, nil
}

func (db *DB) indexVersionFilter(ctx IndexGCContext) hashindex.VersionFilter {
	return func(key []byte, chain []hashindex.Version) []hashindex.Version {
		in := make([]indexVersion, len(chain))
		for i, v := range chain {
			in[i] = fromHashVersion(v)
		}
		out := ctx.FilterVersions(key, in)
		if len(out) == 0 {
			return nil
		}
		versions := make([]hashindex.Version, len(out))
		for i, v := range out {
			versions[i] = toHashVersion(v)
		}
		return versions
	}
}

// IndexArenaStats returns total arena bytes and estimated live bytes.
func (db *DB) IndexArenaStats() (arenaUsed, liveBytes uint64) {
	if db.index == nil || db.index.hash == nil {
		return 0, 0
	}
	stats := db.index.hash.Stats()
	for i := range stats.Shards {
		arenaUsed += stats.Shards[i].ArenaUsed
		liveBytes += stats.Shards[i].LiveBytes
	}
	return arenaUsed, liveBytes
}

// MinActiveSnapshotXmax returns the oldest active snapshot upper bound, or MaxUint64 if none.
func (ctx IndexGCContext) MinActiveSnapshotXmax() uint64 {
	minXmax := uint64(math.MaxUint64)
	for _, r := range ctx.Readers {
		if r.Snapshot.Xmax < minXmax {
			minXmax = r.Snapshot.Xmax
		}
	}
	return minXmax
}

File: rust/crates/turnstone-engine/src/index_gc.rs
// Copyright (c) 2026 Kiruba Sankar Swaminathan
//
// This source code is licensed under the MIT license found in the
// LICENSE file in the root of this source tree.

use std::collections::HashSet;
use std::sync::atomic::Ordering;
use std::time::{SystemTime, UNIX_EPOCH};

use turnstone_hashindex::{Version, VersionFilter};

use crate::types::{IndexVersion, Snapshot, TxStatus};
use crate::Db;

pub const DEFAULT_INDEX_FRAGMENTATION_RATIO: f64 = 3.0;

pub struct IndexGcReader {
    pub snapshot: Snapshot,
    pub my_xid: u64,
    pub update: bool,
}

pub struct IndexGcContext {
    pub readers: Vec<IndexGcReader>,
    pub active_xids: HashSet<u64>,
    pub log_floor: i64,
}

#[derive(Debug, Clone, Default)]
pub struct IndexCompactResult {
    pub shards_compacted: i32,
    pub arena_before: u64,
    pub arena_after: u64,
}

impl Db {
    pub fn build_index_gc_context(&self) -> IndexGcContext {
        let active_txns: Vec<u64> = self.active_txns.lock().values().copied().collect();
        let active_xids: HashSet<u64> = self
            .active_xids
            .lock()
            .keys()
            .copied()
            .collect();
        let readers = Vec::new(); // snapshot readers tracked via active txn tokens in full port
        let _ = active_txns;
        IndexGcContext {
            readers,
            active_xids,
            log_floor: self.scan_floor.load(Ordering::Acquire),
        }
    }

    pub fn maybe_compact_index(&self) -> Result<IndexCompactResult, crate::types::EngineError> {
        let ratio = if self.index_fragmentation_ratio > 0.0 {
            self.index_fragmentation_ratio
        } else {
            DEFAULT_INDEX_FRAGMENTATION_RATIO
        };
        let ctx = self.build_index_gc_context();
        let filter = self.index_wal_retain_filter(&ctx);
        let filter_ref: &VersionFilter<'_> = filter.as_ref();
        let mut before = 0u64;
        let mut after = 0u64;
        let mut compacted = 0i32;
        self.index.with_hash_index(|idx| {
            let n = idx.shard_count();
            for i in 0..n {
                let (arena, live, keys) = idx.filtered_live_bytes(i as i32, Some(filter_ref));
                before += arena;
                if keys == 0 {
                    after += arena;
                    continue;
                }
                if live > 0 && arena as f64 <= live as f64 * ratio {
                    after += arena;
                    continue;
                }
                if let Ok(st) = idx.compact_shard(i as i32, Some(filter_ref)) {
                    compacted += 1;
                    after += st.arena_used;
                }
            }
        });
        if compacted == 0 {
            return Ok(IndexCompactResult::default());
        }
        let res = IndexCompactResult {
            shards_compacted: compacted,
            arena_before: before,
            arena_after: after,
        };
        self.record_index_compact(&res);
        Ok(res)
    }

    fn record_index_compact(&self, res: &IndexCompactResult) {
        if res.shards_compacted <= 0 {
            return;
        }
        self.metrics_hash_shards_compacted
            .fetch_add(res.shards_compacted as u64, Ordering::AcqRel);
        if res.arena_before > res.arena_after {
            self.metrics_hash_compact_reclaimed.fetch_add(
                res.arena_before - res.arena_after,
                Ordering::AcqRel,
            );
        }
        self.metrics_hash_compact_unix.store(
            SystemTime::now()
                .duration_since(UNIX_EPOCH)
                .unwrap_or_default()
                .as_secs() as i64,
            Ordering::Release,
        );
    }

    pub(crate) fn index_wal_retain_filter<'a>(
        &'a self,
        ctx: &'a IndexGcContext,
    ) -> Box<VersionFilter<'a>> {
        let db = self;
        Box::new(move |key: &[u8], chain: &[Version]| {
            let kept = ctx.filter_versions_for_wal_retain(key, chain, db);
            kept.into_iter().map(to_hash_version).collect()
        })
    }

    pub(crate) fn index_version_filter<'a>(
        &'a self,
        ctx: &'a IndexGcContext,
    ) -> Box<VersionFilter<'a>> {
        let db = self;
        Box::new(move |key: &[u8], chain: &[Version]| {
            let in_v: Vec<IndexVersion> = chain.iter().map(from_hash_version).collect();
            let kept = ctx.filter_versions(key, &in_v, db, true);
            kept.into_iter().map(to_hash_version).collect()
        })
    }
}

impl IndexGcContext {
    pub fn filter_versions(
        &self,
        _key: &[u8],
        chain: &[IndexVersion],
        db: &Db,
        apply_log_floor: bool,
    ) -> Vec<IndexVersion> {
        let mask = self.keep_mask(chain, db, apply_log_floor);
        chain
            .iter()
            .zip(mask.iter())
            .filter_map(|(v, k)| if *k { Some(*v) } else { None })
            .collect()
    }

    pub fn filter_versions_for_wal_retain(
        &self,
        key: &[u8],
        chain: &[Version],
        db: &Db,
    ) -> Vec<IndexVersion> {
        let in_v: Vec<IndexVersion> = chain.iter().map(from_hash_version).collect();
        self.filter_versions(key, &in_v, db, false)
    }

    fn keep_mask(&self, chain: &[IndexVersion], db: &Db, apply_log_floor: bool) -> Vec<bool> {
        let n = chain.len();
        let mut keep = vec![false; n];
        if self.readers.is_empty() {
            for (i, v) in chain.iter().enumerate() {
                keep[i] = true;
                if db.clog_status(v.xmin) == TxStatus::Committed {
                    break;
                }
            }
        } else {
            for r in &self.readers {
                for (i, v) in chain.iter().enumerate() {
                    keep[i] = true;
                    if r.update && v.xmin == r.my_xid {
                        break;
                    }
                    if db.is_visible(v.xmin, &r.snapshot) {
                        break;
                    }
                }
            }
        }
        for (i, v) in chain.iter().enumerate() {
            if self.active_xids.contains(&v.xmin) {
                keep[i] = true;
            }
            if apply_log_floor && self.log_floor > 0 && v.offset < self.log_floor {
                keep[i] = false;
            }
        }
        keep
    }
}

fn from_hash_version(v: &Version) -> IndexVersion {
    IndexVersion {
        offset: v.offset,
        value_len: v.value_len,
        xmin: v.xmin,
        tombstone: v.tombstone,
    }
}

fn to_hash_version(v: IndexVersion) -> Version {
    Version {
        offset: v.offset,
        value_len: v.value_len,
        xmin: v.xmin,
        tombstone: v.tombstone,
    }
}
